joint_state_publisher/joint_state_publisher/joint_state_publisher.py

```diff
@@ -269,19 +349,26 @@ def loop(self):
                     joint = self.free_joints[parent]
 
                 if has_position and 'position' in joint:
-                    msg.position[i] = joint['position'] * factor + offset
+                    msg.position[i] = float(joint['position']) * factor + offset
                 if has_velocity and 'velocity' in joint:
-                    msg.velocity[i] = joint['velocity'] * factor
+                    msg.velocity[i] = float(joint['velocity']) * factor
                 if has_effort and 'effort' in joint:
-                    msg.effort[i] = joint['effort']
+                    msg.effort[i] = float(joint['effort'])
 
             if msg.name or msg.position or msg.velocity or msg.effort:
                 # Only publish non-empty messages
                 self.pub.publish(msg)
-            try:
-                r.sleep()
-            except rospy.exceptions.ROSTimeMovedBackwardsException:
-                pass
+
+            # Spin to make sure we service any subscriptions
+            rclpy.spin_once(self.node, timeout_sec=0)
+
+            elapsed = time.time() - start_time
+
+            # TODO(clalancette): Use rclpy.Rate once it is available
```

Comment:
[Rate is available](https://github.com/ros2/rclpy/blob/3b323b564f307cd84d9cb2640c3de8b42ed9295f/rclpy/rclpy/timer.py#L89), but I would definitely recommend a timer callback instead of rate + looping + spin. That would be required to enable this node to be composed.

Reply:
Yeah, excellent call.  Fixed in 2917803.